<a href="https://colab.research.google.com/github/Karthik482/Scania_aps_failure_classification/blob/master/train.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [17]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

**Import test and train data sets**

In [18]:
train = pd.read_csv('aps_failure_training_set.csv')
train.head()

class  aa_000 ab_000      ac_000 ad_000  ...  ee_007  ee_008 ee_009 ef_000 eg_000
0   neg   76698     na  2130706438    280  ...  157956   73224      0      0      0
1   neg   33058     na           0     na  ...   81140   97576   1500      0      0
2   neg   41040     na         228    100  ...  158022   95128    514      0      0
3   neg      12      0          70     66  ...       0       0      0      4     32
4   neg   60874     na        1368    458  ...  311560  433954   1218      0      0

[5 rows x 171 columns]

In [19]:
test=pd.read_csv("aps_failure_test_set.csv")

**DATA CLEANING**

*   converting columns to numeric
*   converting column 1(target) class to binary values





In [ ]:
#converting columns to numeric
train.iloc[:,1:]=train.iloc[:,1:].apply(pd.to_numeric, errors='coerce')

In [ ]:
train.head()

class  aa_000  ab_000  af_000  ...    ee_008  ee_009  ef_000  eg_000
0      0   76698     NaN     0.0  ...   73224.0     0.0     0.0     0.0
1      0   33058     NaN     0.0  ...   97576.0  1500.0     0.0     0.0
2      0   41040     NaN     0.0  ...   95128.0   514.0     0.0     0.0
3      0      12     0.0    10.0  ...       0.0     0.0     4.0    32.0
4      0   60874     NaN     0.0  ...  433954.0  1218.0     0.0     0.0

[5 rows x 132 columns]

In [ ]:
test.iloc[:,1:]=test.iloc[:,1:].apply(pd.to_numeric, errors='coerce')

*Converting target variable into binary*

In [ ]:
class_dict = {'neg':0, 'pos':1}
train["class"] = train["class"].apply(lambda line: class_dict[line])
train_total=train["class"].value_counts()

In [ ]:
test["class"] = test["class"].apply(lambda line: class_dict[line])
test_total=test["class"].value_counts()

Percentage of Target variables

*   Imbalanced target variables will effect our model a lot. We need to see how is more target variables distributed to take correct metrics to evaluate our model




*   Counting percentage of  "zeros"  in each columns (Train)












In [ ]:
total_zero=(train == 0).sum()
p=(total_zero/len(train)*100).sort_values(ascending=False)
p.head(33)


as_000    98.916667
au_000    98.850000
ag_000    98.555000
class     98.333333
ay_009    97.981667
ay_000    97.935000
ag_001    97.645000
ay_001    96.986667
ay_002    96.936667
ay_003    96.796667
az_009    95.741667
cn_000    95.500000
ay_004    95.263333
ef_000    95.035000
dz_000    94.973333
eg_000    94.656667
ea_000    94.225000
cs_009    94.045000
ag_002    93.635000
dj_000    92.993333
dl_000    92.948333
dm_000    92.890000
dk_000    92.791667
ae_000    92.571667
af_000    92.460000
ak_000    92.045000
df_000    91.550000
ar_000    91.546667
dg_000    90.706667
az_008    90.370000
ai_000    89.313333
at_000    88.991667
cn_001    84.106667
dtype: float64

*   Counting percentage of "NaN" in each columns (Train)



In [ ]:
total_na=train.isna().sum()
#percentage=perc
perc=(total_na/len(train)*100).sort_values(ascending=False)
perc.head(8)


br_000    82.106667
bq_000    81.203333
bp_000    79.566667
bo_000    77.221667
cr_000    77.215000
ab_000    77.215000
bn_000    73.348333
bm_000    65.915000
dtype: float64

Counting percentage of "NaN" in each columns (Test)

In [ ]:
total_na=test.isna().sum()
#percentage=perc
perc=(total_na/len(test)*100).sort_values(ascending=False)
perc.head(8)

br_000    82.05625
bq_000    81.13125
bp_000    79.50625
bo_000    77.35000
cr_000    77.26875
ab_000    77.26875
bn_000    73.20625
bm_000    65.91250
dtype: float64

Removing above columns with 65% above NaN value columns from the train and test.

In [ ]:
train_df = train.loc[:, train.isnull().sum() < 0.65*train.shape[0]]
test_df = test.loc[:, test.isnull().sum() < 0.65*test.shape[0]]


*Filling columns with NaN values with mean*

In [ ]:
train_df=train_df.apply(lambda x: x.fillna(x.mean()),axis=0)
test_df=test_df.apply(lambda x:x.fillna(x.mean()),axis=0)

In [ ]:
train_df.head()

class  aa_000        ac_000         ad_000  ...    ee_008  ee_009  ef_000  eg_000
0      0   76698  2.130706e+09     280.000000  ...   73224.0     0.0     0.0     0.0
1      0   33058  0.000000e+00  190620.639314  ...   97576.0  1500.0     0.0     0.0
2      0   41040  2.280000e+02     100.000000  ...   95128.0   514.0     0.0     0.0
3      0      12  7.000000e+01      66.000000  ...       0.0     0.0     4.0    32.0
4      0   60874  1.368000e+03     458.000000  ...  433954.0  1218.0     0.0     0.0

[5 rows x 163 columns]

**Pre Processing data**


1.   spliting target and independent variables in both train and test
2.  feature scaling



In [ ]:
#train target and indepent variables
x_train=train_df.iloc[:,1:]
y_train=train_df.iloc[:,0]
#test target and independent variables
x_test=test_df.iloc[:,1:]
y_test=test_df.iloc[:,0]


In [ ]:
#feature scaling 
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
x_train=sc.fit_transform(x_train)
x_test=sc.transform(x_test)

dtype('float64')

**Data Classification**

*   Naive Bayes
*   XGBoost
*   Random Forest




*Naive Bayes*

In [ ]:
# Training with Naive Bayes
from sklearn.naive_bayes import GaussianNB
classifier=GaussianNB()
classifier.fit(x_train,y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [ ]:
#predecting on test set
y_pred=classifier.predict(x_test)
#comparing predicted results vs actual test results
np.column_stack((y_pred,y_test))

array([[0, 0],
       [0, 0],
       [0, 0],
       ...,
       [0, 0],
       [0, 0],
       [0, 0]])

In [ ]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score
accuracy=accuracy_score(y_test,y_pred)
f1score=f1_score(y_test,y_pred,1)
print('The accuracy is {0}, and f1 score is {1}.'.format(accuracy,f1score))

The accuracy is 0.9685625, and f1 score is 0.5733672603901612.


*XGBoost*

In [ ]:
#traning with gradientboosting
from xgboost import XGBClassifier
classifier=XGBClassifier()
classifier.fit(x_train,y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [ ]:
#predecting on test set
y_pred=classifier.predict(x_test)
#comparing predicted results vs actual test results
np.column_stack((y_pred,y_test))

array([[0, 0],
       [0, 0],
       [0, 0],
       ...,
       [0, 0],
       [0, 0],
       [0, 0]])

In [ ]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import fbeta_score,classification_report
accuracy=accuracy_score(y_test,y_pred)
f1score=fbeta_score(y_test,y_pred,1)
print('The accuracy is {0}, and f1 score is {1}.'.format(accuracy,f1score))

The accuracy is 0.9915, and f1 score is 0.7920489296636085.


*Random Forrest*

In [ ]:
#random forrest
from sklearn.ensemble import RandomForestClassifier
RF=RandomForestClassifier()
RF.fit(x_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [ ]:
#predecting on test set
y_pred=RF.predict(x_test)
#comparing predicted results vs actual test results
np.column_stack((y_pred,y_test))

array([[0, 0],
       [0, 0],
       [0, 0],
       ...,
       [0, 0],
       [0, 0],
       [0, 0]])

In [ ]:
from sklearn.metrics import accuracy_score,precision_score,recall_score
from sklearn.metrics import fbeta_score,classification_report
accuracy=accuracy_score(y_test,y_pred)
f1score=fbeta_score(y_test,y_pred,1)

print('The accuracy is {0}, and f1 score is {1}, precision score is {2},recall score{3}'.format(accuracy,f1score,p,r))
print(classification_report(y_test, y_pred))


The accuracy is 0.9925, and f1 score is 0.8198198198198198, precision score is 0.9381443298969072,recall score0.728
              precision    recall  f1-score   support

           0       0.99      1.00      1.00     15625
           1       0.94      0.73      0.82       375

    accuracy                           0.99     16000
   macro avg       0.97      0.86      0.91     16000
weighted avg       0.99      0.99      0.99     16000

0.9381443298969072


**Implementation: Model Tuning**

> Fine tune the chosen model(XGboost). Use grid search (GridSearchCV) with at least one important parameter tuned with at least 3 different values



In [ ]:
# Import 'GridSearchCV', 'make_scorer'
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer

# Initialize the classifier
classifier = XGBClassifier(random_state=0)

#parameters
parameters={'learning_rate':[0.1,0.05,0.005],
            'max_depth': np.arange(2,10,2),
            'n_estimators':[100,150,200] }

# Make an fbeta_score scoring object using make_scorer()
scorer = make_scorer(fbeta_score, beta=0.5)

# Perform grid search on the classifier using 'scorer' as the scoring method using GridSearchCV()
grid_obj = GridSearchCV(classifier, parameters, scoring=scorer, n_jobs=10, verbose=10)

# Fit the grid search object to the training data and find the optimal parameters using fit()
grid_fit = grid_obj.fit(x_train, y_train)

# Get the estimator
best_clf = grid_fit.best_estimator_

print("Best parameters: ")

# Make predictions using the unoptimized and model
predictions = (classifier.fit(x_train, y_train)).predict(x_test)
best_predictions = best_clf.predict(x_test)

# Report the before-and-afterscores
print("Unoptimized model\n------")
print("Accuracy score on testing data: {:.4f}".format(accuracy_score(y_test, y_pred)))
print("F-score on testing data: {:.4f}".format(fbeta_score(y_test, y_pred, beta = 1)))
print("\nOptimized Model\n------")
print("Final accuracy score on the testing data: {:.4f}".format(accuracy_score(y_test, best_predictions)))
print("Final F-score on the testing data: {:.4f}".format(fbeta_score(y_test, best_predictions, beta = 1 )))



Fitting 5 folds for each of 36 candidates, totalling 180 fits


[Parallel(n_jobs=10)]: Using backend LokyBackend with 10 concurrent workers.
[Parallel(n_jobs=10)]: Done   5 tasks      | elapsed:  1.7min
[Parallel(n_jobs=10)]: Done  12 tasks      | elapsed:  7.9min
[Parallel(n_jobs=10)]: Done  21 tasks      | elapsed: 17.0min
[Parallel(n_jobs=10)]: Done  30 tasks      | elapsed: 24.6min
[Parallel(n_jobs=10)]: Done  41 tasks      | elapsed: 43.5min
[Parallel(n_jobs=10)]: Done  52 tasks      | elapsed: 60.1min
[Parallel(n_jobs=10)]: Done  65 tasks      | elapsed: 69.1min
[Parallel(n_jobs=10)]: Done  78 tasks      | elapsed: 77.1min
[Parallel(n_jobs=10)]: Done  93 tasks      | elapsed: 92.8min
[Parallel(n_jobs=10)]: Done 108 tasks      | elapsed: 113.7min
[Parallel(n_jobs=10)]: Done 125 tasks      | elapsed: 135.5min
[Parallel(n_jobs=10)]: Done 142 tasks      | elapsed: 150.2min
[Parallel(n_jobs=10)]: Done 161 tasks      | elapsed: 175.8min
[Parallel(n_jobs=10)]: Done 180 out of 180 | elapsed: 197.7min remaining:    0.0s
[Parallel(n_jobs=10)]: Done 180

Best parameters: 
Unoptimized model
------
Accuracy score on testing data: 0.9918
F-score on testing data: 0.8000

Optimized Model
------
Final accuracy score on the testing data: 0.9941
Final F-score on the testing data: 0.8609


In [ ]:
print(best_clf)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=8,
              min_child_weight=1, missing=None, n_estimators=200, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)


*Furthur tuning our model with incresing esimators and depth*

In [ ]:
classifier=XGBClassifier(max_depth=10,n_estimators=2000,n_jobs=1)
classifier.fit(x_train,y_train)

In [ ]:
#predecting on test set
y_pred=classifier.predict(x_test)
#comparing predicted results vs actual test results
np.column_stack((y_pred,y_test))

array([[0, 0],
       [0, 0],
       [0, 0],
       ...,
       [0, 0],
       [0, 0],
       [0, 0]])

In [ ]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import fbeta_score,classification_report
accuracy=accuracy_score(y_test,y_pred)
f1score=fbeta_score(y_test,y_pred,1)
print('The accuracy is {0}, and f1 score is {1}.'.format(accuracy,f1score))

The accuracy is 0.9944375, and f1 score is 0.8700729927007298.


Current best tuned model (XGBoost)  f1 score is 0.87. We will perform Random Forrest feature importance on dataset and test our model with selected features again to see if we can improve our model performance

**XGboost Feature importance**

In [20]:
from xgboost import XGBClassifier
classifier=XGBClassifier()
classifier.fit(x_train,y_train)
importance=classifier.feature_importances_
results=[]
for i,v in enumerate(importance):
   if v>0.00001:
    results.append(('Feature:%0d,Score:%.5f'%(i,v)))
results

['Feature:0,Score:0.03530',
 'Feature:3,Score:0.02715',
 'Feature:4,Score:0.02408',
 'Feature:5,Score:0.00909',
 'Feature:6,Score:0.00422',
 'Feature:7,Score:0.00642',
 'Feature:8,Score:0.00732',
 'Feature:9,Score:0.00276',
 'Feature:10,Score:0.00447',
 'Feature:11,Score:0.00266',
 'Feature:12,Score:0.01263',
 'Feature:13,Score:0.00771',
 'Feature:14,Score:0.01108',
 'Feature:15,Score:0.01051',
 'Feature:16,Score:0.01525',
 'Feature:20,Score:0.00453',
 'Feature:22,Score:0.00701',
 'Feature:23,Score:0.00249',
 'Feature:24,Score:0.00702',
 'Feature:25,Score:0.01078',
 'Feature:26,Score:0.01263',
 'Feature:28,Score:0.00426',
 'Feature:29,Score:0.01797',
 'Feature:30,Score:0.00868',
 'Feature:31,Score:0.01109',
 'Feature:32,Score:0.00770',
 'Feature:33,Score:0.00674',
 'Feature:34,Score:0.00474',
 'Feature:35,Score:0.00538',
 'Feature:36,Score:0.00382',
 'Feature:37,Score:0.00624',
 'Feature:39,Score:0.01519',
 'Feature:41,Score:0.00530',
 'Feature:42,Score:0.00412',
 'Feature:43,Score:0.0

In [22]:
#dropping columns from train that has low importance based on above cell results
train=train.drop(train.columns[[3,4,5,15,23,25,26,45,48,49,50,51,53,54,56,57,61,64,66,73,74,75,78,82,86,102,104,106,110,113,114,115,117,119,120,121,122,123,124]],axis=1)
train.head()

class  aa_000 ab_000 af_000 ag_000  ...  ee_007  ee_008 ee_009 ef_000 eg_000
0   neg   76698     na      0      0  ...  157956   73224      0      0      0
1   neg   33058     na      0      0  ...   81140   97576   1500      0      0
2   neg   41040     na      0      0  ...  158022   95128    514      0      0
3   neg      12      0     10      0  ...       0       0      0      4     32
4   neg   60874     na      0      0  ...  311560  433954   1218      0      0

[5 rows x 132 columns]

In [23]:
#dropping columns from test that has low importance based on above cell results
test=test.drop(test.columns[[3,4,5,15,23,25,26,45,48,49,50,51,53,54,56,57,61,64,66,73,74,75,78,82,86,102,104,106,110,113,114,115,117,119,120,121,122,123,124]],axis=1)

*Training model with xgboost feature importance variables*

In [24]:
#converting columns to numeric
train.iloc[:,1:]=train.iloc[:,1:].apply(pd.to_numeric, errors='coerce')
test.iloc[:,1:]=test.iloc[:,1:].apply(pd.to_numeric, errors='coerce')

In [25]:
#converting column (target) class to binary values
class_dict = {'neg':0, 'pos':1}
train["class"] = train["class"].apply(lambda line: class_dict[line])
test["class"] = test["class"].apply(lambda line: class_dict[line])

In [26]:
#Removing columns with 65% above NaN value columns from the train and test.
train_df = train.loc[:, train.isnull().sum() < 0.65*train.shape[0]]
test_df = test.loc[:, test.isnull().sum() < 0.65*test.shape[0]]

In [27]:
#filling na values with mean values
train_df=train_df.apply(lambda x: x.fillna(x.mean()),axis=0)
test_df=test_df.apply(lambda x:x.fillna(x.mean()),axis=0)

In [28]:
#train target and indepent variables
x_train=train_df.iloc[:,1:]
y_train=train_df.iloc[:,0]
#test target and independent variables
x_test=test_df.iloc[:,1:]
y_test=test_df.iloc[:,0]

In [29]:
#feature scaling 
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
x_train=sc.fit_transform(x_train)
x_test=sc.transform(x_test)

In [41]:
classifier=XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=8,
              min_child_weight=1, missing=None, n_estimators=2000, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)
classifier.fit(x_train,y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=8,
              min_child_weight=1, missing=None, n_estimators=2000, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [42]:
#predecting on test set
y_pred=classifier.predict(x_test)
#comparing predicted results vs actual test results
np.column_stack((y_pred,y_test))

array([[0, 0],
       [0, 0],
       [0, 0],
       ...,
       [0, 0],
       [0, 0],
       [0, 0]])

In [43]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import fbeta_score,classification_report
accuracy=accuracy_score(y_test,y_pred)
f1score=fbeta_score(y_test,y_pred,1)
print('The accuracy is {0}, and f1 score is {1}.'.format(accuracy,f1score))
print(classification_report(y_test, y_pred))

The accuracy is 0.9943125, and f1 score is 0.8671532846715329.
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     15625
           1       0.96      0.79      0.87       375

    accuracy                           0.99     16000
   macro avg       0.98      0.90      0.93     16000
weighted avg       0.99      0.99      0.99     16000



**Conclution** 



*We have performed 3 algorithms on our processed dataset and found that XGboost  have higgest F1 score amoung all three 3. Then we finally tuned XGBoost to improve our F1 score which brought the score from 79 to 87%. We further reduced our dataset by removing unimportant features based on XGBoost feature importance. We have applied tunned XGboost classification algorithm on our reduced dataset which has got F1 score to 87%. There was no improvement in F1 score with reduction in data set. We can finally conclude that our best model is xgboost with learning rate 0.5,max_depth=10,n_estimators=2000*